modules/cookbooks/pages/rag.adoc

```diff
@@ -0,0 +1,174 @@
+----
+
+With this pipeline our documents will come out looking something like this:
+
+[source,yaml]
+----
+{
+  "type": "article",
+  "article": {
+    "id": "123foo",
+    "title": "Dogs Stop Barking",
+    "content": "The world was shocked this morning to find that all dogs have stopped barking.",
+    "embeddings": [0.754, 0.19283, 0.231, 0.834], # This vector will actually have 768 dimensions
+  }
+}
+----
+
+Now we can send that to Postgres using the xref:components:outputs/sql_insert.adoc[`sql_insert`] output. We'll also take advantage of the `init_statement` functionality to setup our `pgvector` and our table.
+
+[source,yaml]
+----
+output:
+  sql_insert:
+    driver: postgres
+    dsn: "${PG_CONNECTION_STRING}"
+    init_statement: |
+      CREATE EXTENSION IF NOT EXISTS vector;
+      CREATE TABLE IF NOT EXISTS searchable_text (
+        id varchar(128) PRIMARY KEY,
+        title text NOT NULL,
+        body text NOT NULL,
+        embeddings vector(768) NOT NULL
+      );
+      CREATE INDEX IF NOT EXISTS text_hnsw_index
+        ON searchable_text 
+        USING hnsw (embeddings vector_l2_ops);
+    table: searchable_text
+    columns: ["id", "title", "body", "embeddings"]
+    args_mapping: "[this.article.id, this.article.title, this.article.content, this.article.embeddings.vector()]"
+----
+
+We can save this pipeline as `indexing.yaml` and run it with `rpk connect run ./indexing.yaml` to test and make sure our postgres table is being populated with embeddings.
+
+== Generating search responses
+
+In order to generate responses to questions over our dataset we're creating embeddings for, we'll use an xref:components:inputs/http_server.adoc[HTTP server input] to recieve the questions. Since we'll be running this pipeline in Streams mode, our HTTP server will xref:guides:streams_mode/about.adoc#http-endpoints[endpoints prefixed with stream identifier], which in our case is the same as the filename: `search`.
+
+[source,yaml]
+----
+input:
+  http_server:
+    path: /
+    allowed_verbs: [ GET ]
+    sync_response:
+      headers:
+        Content-Type: application/json
+----
+
+Our user query will come in as the query parameter `q` like so `http://localhost:4195/search?q=question_here`. Since query parameters as exposed as metadata in the `http_server` input, we can reference that in bloblang with `@q`.
+
+[source,yaml]
+----
+pipeline:
+  processors:
+    - label: compute_embeddings
+      ollama_embeddings:
+        model: nomic-embed-text
+        text: "search_query: %s".format(@q)
+----
+
+Now our payload is the embeddings vector and we can use Postgres to fetch the top 3 most similar documents to our embeddings vector using the xref:components:processors/sql_raw.adoc[`sql_raw`] processor.
+
+[source,yaml]
+----
+    - sql_raw:
+        driver: "postgres"
+        dsn: "${PG_CONNECTION_STRING}"
+        query: SELECT title, body FROM searchable_text ORDER BY embeddings <-> $1 LIMIT 3
```

Comment:
The <-> notation is showing as an editable field in the preview. You can switch off this functionality in the snippet by adding this inside the start of the code block. [,yaml, role="no-placeholders"].